### ISSUE TITLE
Plugin errors when trying to load plugins without package.json files (gatsby-plugin-image)

### ISSUE THREAD

**Issue by rramphal** [Js-Brecht/gatsby-plugin-ts-config#23]
When using [`gatsby-plugin-image`](https://www.npmjs.com/package/gatsby-plugin-image) with `gatsby-plugin-ts-config`, we get the following error on `gatsby develop`:

```
There was a problem loading plugin "gatsby-plugin-image". Perhaps you need to install its package?
```

Running `gatsby develop --verbose` errors out with a bit more information:

```
plugin "gatsby-plugin-image" threw the following error:
 ENOENT: no such file or directory, open 'node_modules/gatsby-plugin-image/dist/package.json'
```

Indeed, that file does not exist on the filesystem even though the `gatsby-plugin-image` module is present within `package.json` and the module is installed. However, converting `gatsby-config.ts` back to `gatsby-config.js` allows build to complete successfully without any further changes.

**Comment by rramphal** [comment]
Also posted https://github.com/gatsbyjs/gatsby/discussions/27950#discussioncomment-154415 for potentially more information.

**Comment by Js-Brecht** [comment]
It's because Gatsby looks for the `package.json` next to `resolve`... if `resolve` isn't actually resolved yet, it does that.

The problem here is that I have to resolve plugins before sending them to Gatsby, otherwise it will fail to pick them up when using pnpm or yarn2.  The `main` for `gatsby-plugin-image` is under `/dist`, instead of under `/`, so `resolve` is going to point to `/dist/gatsby-image.js` instead of `/`, so Gatsby looks for `/dist/package.json` instead of `/package.json`.

The solution for this is probably going to be utilizing the [`parentDir` field](https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby/src/bootstrap/load-themes/index.js#L147) on the plugins array, but the core will need to be updated to not overwrite that field if it already exists.  That field is what makes it possible for Gatsby to resolve plugins that belong to themes, which is pretty much what this plugin is, and is the entire reason the process works like it currently does.

**Comment by Js-Brecht** [comment]
@rramphal I imagine you will get the same error if you did something like this:
```js
// gatsby-node.js
module.exports = {
  plugins: [
    {
      resolve: require.resolve("gatsby-plugin-image"),
    }
  ]
}
```

Can you confirm?

**Comment by rramphal** [comment]
Hey @Js-Brecht, thank you for jumping on this so quickly! I do hope that the `gatsby` team gets to your PR to unblock a fix.

Also, yes, I can confirm that the code you suggested also errors in the same way.

**Comment by ascorbic** [comment]
I'm a bit confused here. Why can't you do `require.resolve('gatsby-plugin-image/package.json', {paths: [program.directory]})`?

**Comment by rramphal** [comment]
@ascorbic Sorry, I'm not familiar with the `program.directory` bit. Where is programs defined in a standard `gatsby-config` file?

**Comment by ascorbic** [comment]
Oh, sorry that was addressed to @Js-Brecht. I meant in the plugin not in gatsby-config. It's just the project root. Plugins can get it from the redux store.

**Comment by Js-Brecht** [comment]
> I'm a bit confused here. Why can't you do `require.resolve('gatsby-plugin-image/package.json', {paths: [program.directory]})`?

This plugin does do something very similar to that already.  Only difference is that it uses a scoped require much like the core does, so that it is compatible with yarn2.

The path resolution this plugin does is why this problem has come up in the first place.  It was already something I had wanted to fix, because, as I mentioned, it has other problems with it too.

Some work could [probably done here, as well](https://github.com/gatsbyjs/gatsby/blob/721f42aa210943a5cfe9b08fe93ad31ae315c447/packages/gatsby/src/bootstrap/load-plugins/load.ts#L93-L112), so that the core will support absolute paths better.  The alternative is moving the `main` for `gatsby-plugin-image` into the root of its package

**Comment by ascorbic** [comment]
The difference I mean is resolving 'gatsby-plugin-image/package.json' rather than resolving the root and then trying to resolve the package.json relaltive to that.

**Comment by Js-Brecht** [comment]
In what context?  In this plugin?

**Comment by Js-Brecht** [comment]
That would probably work as a patch, but it still leaves problems with deduplication.

**Comment by ascorbic** [comment]
Yes, though looking at your code I realise that is what you are doing. So if I'm understanding it, what's happening is that your plugin is resolving the plugin name into an absolute path, but the problem is that path is to the entry point directory, rather than the plugin root. Is that something that's required? Can you resolve it to the package.json dirname and send that to gatsby instead? Or am I misunderstanding?

**Comment by Js-Brecht** [comment]
It resolves to the plugin’s root directory [here](https://github.com/Js-Brecht/gatsby-plugin-ts-config/blob/master/src/utils/endpoints.ts#L131-L133), by getting the dirname of `<plugin_name>/package.json`.

After considering your suggestion, however, I believe that It would work if I were to _not_ get the dirname for it, leaving it resolved to `/package.json`.  Gatsby would get the `dirname` for it, and then tack on `/package.json` when reading it.

That would be a quick fix that I could publish immediately.  Not something I’d want to leave hanging around forever, though.  Not really a fan of breaking the deduplication, especially where a user doesn’t have control over it.  I’m a bit surprised that it hasn’t created more problems already, but I’m sure it’s only a matter of time

**Comment by Js-Brecht** [comment]
About why this issue is occurring: it is actually _because_ it has been resolved to the plugin root directory.

Where the error is actually being thrown is [here](https://github.com/gatsbyjs/gatsby/blob/721f42aa210943a5cfe9b08fe93ad31ae315c447/packages/gatsby/src/bootstrap/load-plugins/load.ts#L100-L112).  Gatsby sees that it is an absolute path already, and the `require.resolve()` on the plugin's root directory is resolving the entry point.  Then it gets the dirname of that (in this case, `/dist`), and tacks on `package.json` for reading... but `/dist/package.json` isn't valid

**Comment by Js-Brecht** [comment]
@rramphal I've gone ahead and merged that quick-fix: `gatsby-plugin-ts-config@1.1.5` should support `gatsby-plugin-image`.  The rest of it will have to wait.  Please let me know if you continue having trouble

**Comment by rramphal** [comment]
@Js-Brecht I just tested and it works! Thank you so much for your insanely quick turnaround on this. 🙏🏾
